neuralmonkey/model/model_part.py

```diff
@@ -1,147 +1,114 @@
 """Basic functionality of all model parts."""
-
 from abc import ABCMeta
-from contextlib import contextmanager
-from typing import Any, Callable, Dict, List, Set, Tuple
+from typing import MutableSet, Set, List, Tuple
+
+from neuralmonkey.model.parameterized import Parameterized, InitializerSpecs
+from neuralmonkey.model.feedable import Feedable
+
+
+class GenericModelPart(metaclass=ABCMeta):
+    """Base class for Neural Monkey model parts.
+
+    Neural Monkey dynamically decides which model parts are in use when using a
+    specific trainer or a runner. Each trainer/runner holds a reference to a
+    top-level model part, which is then responsible for collecting references
+    to all `Parameterized` and `Feedable` objects that contribute to the
+    computation of its Tensors. This behavior is implemented using the
+    `get_dependencies` method, which is called recursively on all instances of
+    `GenericModelPart` class that are references from within a model part.
 
-import tensorflow as tf
+    Apart from the `get_dependencies` method, this class also provides two
+    properties: `list_dependencies` and `singleton_dependencies` which store
+    the names of the Python class attributes that are regarded as potential
+    dependents of the `GenericModelPart` object. These dependents are
+    automatically checked for type and when they are instances of the
+    `GenericModelPart` class, results of their `get_dependencies` are united
+    and returned as dependencies of the current object.
+    """
+
+    @property
+    def list_dependencies(self) -> List[str]:
+        """Return a list of attributes regarded as lists of dependents."""
+        return ["attentions", "encoders"]
+
+    @property
+    def singleton_dependencies(self) -> List[str]:
```

Comment:
Název singleton IMHO evokuje ten Javařskej návrhovej vzor, tak by to chtělo lepší název, i kdyby to zůstalo.

Reply:
vyhodil jsem to a zůstalo jen property dependencies